### PR TITLE
Implement Feature Adversaries in PyTorch

### LINKED ISSUE #949: Implement Feature Adversaries attack framework-specific in PyTorch and TensorFlow
The Feature Adversaries attack is an important attack for robustness evaluations and is available in ART in a framework-agnostic implementation in `art.attacks.evasion.FeatureAdversaries`. Because of its importance we would like to propose the implementation for framework-specific version in TensorFlow and PyTorch to take advantage of improved performance provided by these frameworks.

Paper link: https://arxiv.org/abs/1511.05122

### PR DESCRIPTION
**Pull request by hesseltuinhof**
# Description

The PR offers an additional port of the Feature Adversaries attack from NumPy into PyTorch. Furthermore, improved the Numpy implementation by adding some additional checks on input and output. A simple notebook demonstrating adversarial features in PyTorch has been created.

Fixes #949 

## Type of change

Please check all relevant options.

- [x] New feature (non-breaking)

# Testing

**Test Configuration**:
- Fedora 33
- Python 3.8.10
- ART 1.7 development
- PyTorch 1.8.1+cpu

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by hesseltuinhof** [comment]
The PR has been cleaned up and rebased. Thanks, @beat-buesser for your contributions!

File: art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        "step_size",
+        "random_start",
+        "verbose",
+    ]
+
+    _estimator_requirements = (BaseEstimator, NeuralNetworkMixin)
+
+    def __init__(
+        self,
+        classifier: "CLASSIFIER_NEURALNETWORK_TYPE",
```
Comment (beat-buesser) [inline art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py:66]:
```suggestion
        estimator: "PyTorchEstimator",
```

File: art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        max_iter: int = 100,
+        batch_size: int = 32,
+        step_size: Optional[Union[int, float]] = None,
+        random_start: bool = False,
+        verbose: bool = True,
+    ):
+        """
+        Create a :class:`.FeatureAdversariesPyTorch` instance.
+
+        :param classifier: A trained classifier.
```
Comment (beat-buesser) [inline art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py:81]:
```suggestion
        :param estimator: A trained estimator.
```

File: art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        :param delta: The maximum deviation between source and guide images.
+        :param lambda_: Regularization parameter of the L-inf soft constraint.
+        :param layer: Index of the representation layer.
+        :param max_iter: The maximum number of iterations.
+        :param batch_size: Batch size.
+        :param step_size: Step size for PGD optimizer.
+        :param random_start: Randomly initialize perturbations, when using Projected Gradient Descent variant.
+        :param verbose: Show progress bars.
+        """
+        super().__init__(estimator=classifier)
```
Comment (beat-buesser) [inline art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py:94]:
```suggestion
        super().__init__(estimator=estimator)
```

File: art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+                # pgd step
+                perturbation = -adv.grad.data.sign() * self.step_size
+                adv = adv + perturbation
+                perturbation = torch.clamp(adv.data - x.data, -self.delta, self.delta)
+                adv = torch.autograd.Variable(
+                    torch.clamp(x.data + perturbation, self.estimator.clip_values[0], self.estimator.clip_values[1]),
+                    requires_grad=True,
+                )
+        else:
+            # optimize soft constraint problem with chosen optimzier
```
Comment (beat-buesser) [inline art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py:156]:
```suggestion
            # optimize soft constraint problem with chosen optimizer
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @hesseltuinhof Thank you very much for implementing the Feature Adversaries evasion attack in PyTorch and for fixing bugs of other tools along the way! This new attack will find a lot of applications.

Your pull request is very well prepared! I have made a few suggestions to relax typing on the attacked estimator, basically the attack can attack all estimators created with PyTorch. I think the title of cell 10 of the notebook should change Unconstraint to Unconstrained.

File: art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py (hunk 1 above)
Comment (hesseltuinhof) [reply art/attacks/evasion/feature_adversaries/feature_adversaries_pytorch.py:66]: As just discussed, I added the additional `PYTORCH_ESTIMATOR_TYPE` to `art/utils.py` and using this one now :+1:

**Comment by hesseltuinhof** [comment]
Thanks, @beat-buesser for the review! I attended all the comments and rebased the PR accordingly :)

**Review by beat-buesser — APPROVED** [review]
